.github/workflows/test.yml

```diff
@@ -3,41 +3,137 @@ name: CI
 on: [push, pull_request]
 
 jobs:
-    unix-test:
+    build-test:
         runs-on: ${{ matrix.os }}
         strategy:
             fail-fast: false
             matrix:
                 os: [ubuntu-latest, macos-latest]
+                build: [general, webrender]
 
         steps:
             - uses: actions/checkout@v2
 
+            - name: rust setup
+              id: rust-setup
+              run: |
+                  rustup install $(cat rust-toolchain)
+                  rustup component add rustfmt-preview
+                  echo "::set-output name=week::$(/bin/date -u '+%Yy_%mm_%Ww')"
+                  echo "::set-output name=date::$(/bin/date)"
+
+            - name: rust cache
+              uses: actions/cache@v2
+              with:
+                  path: |
+                      ~/.cargo/registry/cache
+                      ~/.cargo/git
+                      ./rust_src/target
+                  key: ${{ runner.os }}-${{ matrix.build }}-rust-${{ hashFiles('**/rust-toolchain', '**/Cargo.toml*', '**/rustfmt.toml') }}-${{ steps.rust-setup.outputs.week }}
```

Comment:
Why don't we use `${{ runner.os }}-cargo-${{ hashFiles('**/rust-toolchain', '**/Cargo.lock') }}` instead?
The `Cargo.toml` can be hand edited (`Cargo.lock` is better as key) and `rustfmt.toml` is just a format file.
Also the `-${{ steps.rust-setup.outputs.week }}` is not needed as we use `rust-toolchain already`

Reply:
Here is my PR that's using the proposed key, #216 
Hope that we can use the same key so the cache will be hit for both.

Reply:
Don't we also define build options like codegen-unit in file `Cargo.toml`/`Cargo.toml.in`?
`-${{ steps.rust-setup.outputs.week }}`, `restore-keys updates` cache every week.
If features are added without changing dependencies, cache is broken.

Reply:
> Don't we also define build options like codegen-unit in file `Cargo.toml`/`Cargo.toml.in`?

That would be reflected in `Cargo.lock` wouldn't it? Which contains all of the dependencies packages the will be built.


> `-${{ steps.rust-setup.outputs.week }}`, `restore-keys updates` cache every week.
> If features are added without changing dependencies or fix build option, cache is broken.

I'm not sure if I get this correctly but you mean adding the `outputs.week` there is to invalidate the cache weekly?
Is that needed? Especially since the `restore-keys` just cut off that part brutally anyway.

Reply:
> That would be reflected in `Cargo.lock` wouldn't it? Which contains all of the dependencies packages the will be built.

I couldn't find next options at lockfile.
```toml
[profile.dev]
panic = "abort"

[profile.release]
panic = "abort"
codegen-units = 1
lto = true
# This reduces the binary size from 125M to 70M due
# to Deno inclusion
opt-level = 'z' # Optimize for size
```
There is only dependency information.
After changing the profile option, considered hit rate.
I just thought defensively and I don't have a strong opinion as changes to file `Cargo.toml` are not frequent.

> I'm not sure if I get this correctly but you mean adding the `outputs.week` there is to invalidate the cache weekly?

It is not completely invalidated.
Build using old cache values, then upload updated cache(`1~2m`).
This is a policy to maintain freshness while making the most of the benefits of existing caches.

Reply:
> There is only dependency information.
> After changing the profile option, considered hit rate.
> I just thought defensively and I don't have a strong opinion as changes to file `Cargo.toml` are not frequent.

Makes sense. However, for the caching stuff, this would be the same if you making a small change and rebuild in the `emacs-ng`, wouldn't it. So I expect the compiler should be able to detect that the cache of compiled binary is not usable in that case anyway.
The problem here is the cache of `~/.cargo/...`, which is just downloads of the prebuilt crate and source code. With the new cache key is created for changes in `Cargo.toml`, the cache there will be compressed, duplicated, and cached again, which's taking storage.



> It is not completely invalidated.
> Build using old cache values, then upload updated cache(`1~2m`).
> This is a policy to maintain freshness while making the most of the benefits of existing caches.

So after a week, a fresh cache will be created and taking more cache storage?
The updated cache is diff from the previous cache?

Reply:
That's a good point.

**Compile Option Changes**
I intentionally reduced the debug information and compared the cache size.
https://github.com/black7375/emacs-ng/actions/runs/740067338
https://github.com/black7375/emacs-ng/actions/runs/740126744
- old cache: `~1762 MB`(Ubuntu General), `~1864 MB`(Ubuntu Webrender), `~1674 MB`(Mac General)
- new cache: `~1611 MB`(Ubuntu General), `~1689 MB`(Ubuntu Webrender), `~1629 MB`(Mac General)

**Lock File Changes**
After run `cargo update` to `Cargo.lock`
There may be slight increases.

- old cache: `~1762 MB`(Ubuntu General), `~1864 MB`(Ubuntu Webrender), `~1674 MB`(Mac General)
- new cache: ``(Ubuntu General), ``(Ubuntu Webrender), ``(Mac General)

...testing

Reply:
But you can solve it like this.

https://github.com/matthiaskrgr/cargo-cache
`cargo cache --autoclean ` or `cargo cache --autoclean-expensive `

How about running after installing only when it is a new key?
It can be possible with `ouput.cache-hit`
```
if: steps.rust-cache.outputs.cache-hit != 'true'
```